File: ch_erp15/ch_erp15/custom/purchase_invoice.py
from frappe.utils import flt, rounded, money_in_words, fmt_money
from erpnext.accounts.doctype.purchase_invoice.purchase_invoice import PurchaseInvoice


class CustomPurchaseInvoice(PurchaseInvoice):
    def validate(self):
        super().validate()

        if self.custom_purchase_type != "Marginal":
            return

        total_margin_taxable = 0
        total_gst = 0
        total_amount = 0

        for item in self.items:
            qty = flt(item.qty)
            rate = flt(item.rate)
            margin_unit = flt(item.custom_unit_taxable_value)

            item_amount = qty * rate
            margin_taxable = qty * margin_unit

            item.amount = item_amount
            item.base_amount = item_amount

            total_amount += item_amount
            total_margin_taxable += margin_taxable

        for tax in self.taxes:
            tax.included_in_print_rate = 1
            tax.tax_amount = 0

            if tax.charge_type == "On Net Total":
                tax.tax_amount = flt(
                    (total_margin_taxable * flt(tax.rate)) / 100,
                    tax.precision("tax_amount"),
                )
                total_gst += tax.tax_amount

            tax.base_tax_amount = tax.tax_amount
            tax.tax_amount_after_discount_amount = tax.tax_amount
            tax.base_tax_amount_after_discount_amount = tax.tax_amount
            tax.base_total = tax.tax_amount
            tax.total = tax.tax_amount

        net_total = total_amount - total_gst

        self.total = total_amount
        self.base_total = total_amount
        self.net_total = net_total
        self.base_net_total = net_total

        self.total_taxes_and_charges = total_gst
        self.base_total_taxes_and_charges = total_gst
        self.taxes_and_charges_added = total_gst

        self.grand_total = total_amount
        self.base_grand_total = total_amount
        self.rounded_total = total_amount
        self.base_rounded_total = total_amount
        self.outstanding_amount = total_amount

        self.rounding_adjustment = 0
        self.base_rounding_adjustment = 0

        self.in_words = money_in_words(self.rounded_total, self.currency)
        self.base_in_words = money_in_words(self.rounded_total, self.currency)

        self.set_custom_tax_breakup()

    def get_gl_entries(self, warehouse_account=None):
        """Build custom GL entries for Marginal purchase type"""

        if self.custom_purchase_type != "Marginal":
            return super().get_gl_entries(warehouse_account)

        gl_entries = []

        total_margin_taxable = sum(
            flt(item.qty) * flt(item.custom_unit_taxable_value)
            for item in self.items
        )

        grand_total = flt(self.grand_total)

        total_tax = 0
        tax_entries = []
        for tax in self.taxes:
            tax_amt = 0
            if tax.charge_type == "On Net Total" and flt(tax.rate) > 0:
                tax_amt = flt(
                    (total_margin_taxable * flt(tax.rate)) / 100,
                    tax.precision("tax_amount"),
                )
            total_tax += tax_amt
            if tax_amt:
                tax_entries.append({
                    "account": tax.account_head,
                    "amount": tax_amt,
                    "cost_center": tax.cost_center or self.cost_center,
                })

        expense_total = flt(grand_total - total_tax)

        # Against accounts string
        against_accounts = list(set(
            [item.expense_account for item in self.items]
            + [t["account"] for t in tax_entries]
        ))

        gl_entries.append(
            self.get_gl_dict(
                {
                    "account": self.credit_to,
                    "party_type": "Supplier",
                    "party": self.supplier,
                    "against": ", ".join(against_accounts),
                    "credit": grand_total,
                    "credit_in_account_currency": grand_total,
                    "against_voucher": (
                        self.return_against
                        if self.get("is_return") and self.return_against
                        else self.name
                    ),
                    "against_voucher_type": self.doctype,
                    "cost_center": self.cost_center,
                    "project": self.project,
                },
                self.party_account_currency,
                item=self,
            )
        )

        total_full = sum(flt(item.qty) * flt(item.rate) for item in self.items)
        running_debit = 0

        for idx, item in enumerate(self.items):
            item_full = flt(item.qty) * flt(item.rate)

            # Last item absorbs any rounding difference
            if idx == len(self.items) - 1:
                item_debit = flt(expense_total - running_debit)
            else:
                proportion = item_full / total_full if total_full else 0
                item_debit = flt(
                    expense_total * proportion,
                    item.precision("base_net_amount"),
                )

            running_debit += item_debit

            gl_entries.append(
                self.get_gl_dict(
                    {
                        "account": item.expense_account,
                        "against": self.credit_to,
                        "debit": item_debit,
                        "debit_in_account_currency": item_debit,
                        "cost_center": item.cost_center or self.cost_center,
                        "project": item.project or self.project,
                    },
                    item=item,
                )
            )

        for t in tax_entries:
            gl_entries.append(
                self.get_gl_dict({
                    "account": t["account"],
                    "against": self.credit_to,
                    "debit": t["amount"],
                    "debit_in_account_currency": t["amount"],
                    "cost_center": t["cost_center"],
                })
            )

        return gl_entries


    def set_custom_tax_breakup(self):
        try:
            headers = ["HSN/SAC", "Taxable Amount"] + [
                tax.description or tax.account_head or "Tax" for tax in self.taxes]
 
            total_taxable = sum(flt(item.qty) * flt(item.get("custom_unit_taxable_value"))
                for item in self.items)
 
            # group hsn
            hsn_map = {}
            for item in self.items:
                taxable_value = flt(item.qty) * flt(item.get("custom_unit_taxable_value"))
                amount = flt(item.amount) or flt(item.base_amount)
                if taxable_value <= 0:
                    continue
 
                hsn = item.get("gst_hsn_code") or ""
                if hsn not in hsn_map:
                    hsn_map[hsn] = {"amount": 0, "taxable_value": 0}
                hsn_map[hsn]["amount"] += amount
                hsn_map[hsn]["taxable_value"] += taxable_value
 
            rows = ""
            for hsn, data in hsn_map.items():
                proportion = data["taxable_value"] / total_taxable if total_taxable else 0
 
                cells = f"<td>{hsn}</td>"
                cells += f"<td style='text-align:right'>{fmt_money(data['amount'], currency='INR')}</td>"
                for tax in self.taxes:
                    item_tax = flt(flt(tax.tax_amount) * proportion)
                    cells += (
                        f"<td style='text-align:right'>"
                        f"({flt(tax.rate)}%) {fmt_money(item_tax, currency='INR')}"
                        f"</td>"
                    )
                rows += f"<tr>{cells}</tr>"
 
            if not rows:
                rows = f"<tr><td colspan='{len(headers)}' style='text-align:center'>No tax data</td></tr>"
 
            header_html = f"<th>{headers[0]}</th>"
            header_html += "".join(f"<th style='text-align:right'>{h}</th>" for h in headers[1:])
            self.other_charges_calculation = f"""
            <table class="table table-bordered">
                <thead><tr>{header_html}</tr></thead>
                <tbody>{rows}</tbody>
            </table>
            """
 
        except Exception:
            import frappe
            frappe.log_error(title="Tax Breakup Error", message=frappe.get_traceback())
            self.other_charges_calculation = ""

File: ch_erp15/ch_erp15/custom/purchase_invoice.js

frappe.ui.form.on("Purchase Invoice", {

    onload(frm) {
        frm.cscript = frm.cscript || {};
        if (!frm._original_calculation) {
            frm._original_calculation = frm.cscript.calculate_taxes_and_totals;
        }
        frm.cscript.calculate_taxes_and_totals = function() {
            if (frm.doc.custom_purchase_type === "Marginal") {
                apply_zero_tax(frm);
                apply_marginal_scheme(frm);
                apply_on_marginal(frm);
                apply_taxable_scheme(frm);

                return;
            }
            if (frm._original_calculation) {
                frm._original_calculation.call(frm);
            }
        };
    },
    refresh(frm) {
        apply_zero_tax(frm);
        apply_marginal_scheme(frm);
        apply_on_marginal(frm);
        apply_taxable_scheme(frm);

    },

    custom_purchase_type(frm) {
        apply_zero_tax(frm);
        apply_marginal_scheme(frm);
        apply_on_marginal(frm)
        apply_taxable_scheme(frm)

    },

    validate(frm) {
        apply_marginal_scheme(frm);
        apply_on_marginal(frm);
        apply_zero_tax(frm);
        apply_taxable_scheme(frm)

        
        
    }
});

function apply_on_marginal(frm) {
    const is_m = frm.doc.custom_purchase_type === "Marginal";
    const style_id = "mt-hide-style";
    let css = "";
 

 
    let old = document.getElementById(style_id);
    if (old) old.remove();
 
    let style = document.createElement("style");
    style.id = style_id;
    style.innerHTML = css;
    document.head.appendChild(style);
 
    frm.refresh_field("items");
}
 





frappe.ui.form.on("Purchase Invoice Item", {

    qty(frm) {
        apply_marginal_scheme(frm);
        apply_zero_tax(frm);
        apply_taxable_scheme(frm)


        
    },

    rate(frm) {
        apply_marginal_scheme(frm);
        apply_zero_tax(frm);
        apply_taxable_scheme(frm)


    },

    custom_unit_taxable_value(frm) {
        apply_marginal_scheme(frm);
        apply_zero_tax(frm);
        apply_taxable_scheme(frm)

        

    }
});

frappe.ui.form.on("Purchase Taxes and Charges", {

    rate(frm) {
        apply_marginal_scheme(frm);
        apply_zero_tax(frm);
        apply_taxable_scheme(frm)

    }
});










function apply_zero_tax(frm) {

    if (frm.doc.custom_purchase_type !== "Unregistered") return;

    let total_amount = 0;


    
    (frm.doc.items || []).forEach(item => {

        let qty = flt(item.qty);
        let rate = flt(item.rate);

        let amount = qty * rate;

        item.amount = amount;
        item.base_amount = amount;

        total_amount += amount;
    });

    (frm.doc.taxes || []).forEach(row => {

        // row.rate = 0;

        row.tax_amount = 0;
        row.base_tax_amount = 0;

        row.total = total_amount;
        row.base_total = total_amount;
    });

    frm.doc.net_total = total_amount;
    frm.doc.base_net_total = total_amount;

    frm.doc.total = total_amount;
    frm.doc.base_total = total_amount;

    frm.doc.total_taxes_and_charges = 0;
    frm.doc.base_total_taxes_and_charges = 0;

    frm.doc.taxes_and_charges_added = 0;
    frm.doc.base_taxes_and_charges_added = 0;

    frm.doc.grand_total = total_amount;
    frm.doc.outstanding_amount= total_amount;
    frm.doc.base_grand_total = total_amount;

    frm.doc.rounded_total = Math.round(total_amount);
    frm.doc.base_rounded_total = Math.round(total_amount);

    frm.refresh_fields([
        "items",
        "taxes",
        "net_total",
        "grand_total",
        "rounded_total"
    ]);
}













function apply_marginal_scheme(frm) {

    if (frm.doc.custom_purchase_type !== "Marginal") return;

    let total_margin_taxable = 0;
    let total_gst = 0;
    let total_exempted = 0;

    (frm.doc.items || []).forEach(item => {

        let qty = flt(item.qty);
        let rate = flt(item.rate);
        let margin_unit = flt(item.custom_unit_taxable_value);

        if (qty <= 0 || rate <= 0 || margin_unit <= 0) return;

        let item_amount = qty * rate;
        let margin_taxable = qty * margin_unit;

        item.amount = item_amount;
        item.base_amount = item_amount;
        item.taxable_value = margin_taxable;

        total_margin_taxable += margin_taxable;
    });

    frm.refresh_field("items");

    (frm.doc.taxes || []).forEach(tax => {

        tax.tax_amount = 0;
        tax.amount = 0;
        tax.total = 0;

        if (tax.charge_type === "On Net Total") {

            let tax_amount = (total_margin_taxable * flt(tax.rate)) / 100;

            tax.tax_amount = tax_amount;
            tax.amount = tax_amount;
            tax.total = tax_amount;

            tax.base_tax_amount = tax_amount;
            tax.base_amount = tax_amount;

            total_gst += tax_amount;
        }

    });

    frm.refresh_field("taxes");

    (frm.doc.items || []).forEach(item => {

        let qty = flt(item.qty);
        let margin_unit = flt(item.custom_unit_taxable_value);

        if (qty <= 0 || flt(item.amount) <= 0) return;

        let margin_taxable = qty * margin_unit;

        let item_gst = 0;
        if (total_margin_taxable > 0) {
            item_gst = (margin_taxable / total_margin_taxable) * total_gst;
        }

        let exempted_value = flt(item.amount) - margin_taxable - item_gst;

        if (exempted_value < 0) {
            exempted_value = 0;
        }

        item.custom_exempted_value = exempted_value;

        total_exempted += exempted_value;
    });

    frm.refresh_field("items");


    frm.doc.net_total = total_margin_taxable;
    frm.doc.base_net_total = total_margin_taxable;

    frm.doc.taxes_and_charges_added = total_gst;
    frm.doc.base_taxes_and_charges_added = total_gst;

    frm.doc.total_taxes_and_charges = total_gst;
    frm.doc.base_total_taxes_and_charges = total_gst;

    let custom_grand_total = total_margin_taxable + total_gst + total_exempted;

    frm.doc.grand_total = custom_grand_total;
    frm.doc.base_grand_total = custom_grand_total;
    frm.doc.rounded_total = Math.round(custom_grand_total);
    frm.doc.outstanding_amount = custom_grand_total;

    frm.refresh_fields();
}



function apply_taxable_scheme(frm) {

    if (frm.doc.custom_purchase_type !== "Taxable") return;

    let total_taxable = 0;
    let total_gst = 0;

    
    (frm.doc.items || []).forEach(item => {

        let qty = flt(item.qty);
        let unit_taxable = flt(item.custom_unit_taxable_value);

        if (qty <= 0 || unit_taxable <= 0) {
            item.rate = 0;
            item.amount = 0;
            item.base_amount = 0;
            item.taxable_value = 0;
            return;
        }

        let rate = unit_taxable * 1.18;
        let amount = qty * rate;
        let taxable_value = qty * unit_taxable;

        item.rate = rate;
        item.amount = amount;
        item.base_amount = amount;
        item.taxable_value = taxable_value;

        total_taxable += taxable_value;
    });

    frm.refresh_field("items");

    
    total_gst = total_taxable * 0.18;

    (frm.doc.taxes || []).forEach(tax => {

        tax.tax_amount = total_gst;
        tax.amount = total_gst;
        tax.total = total_gst;

        tax.base_tax_amount = total_gst;
        tax.base_amount = total_gst;
    });

    frm.refresh_field("taxes");
    let gross_total = total_taxable + total_gst;
    let discount = flt(frm.doc.discount_amount);

    let final_total = gross_total - discount;

    if (final_total < 0) final_total = 0;

    frm.doc.net_total = total_taxable;
    frm.doc.base_net_total = total_taxable;

    frm.doc.total_taxes_and_charges = total_gst;
    frm.doc.base_total_taxes_and_charges = total_gst;

    frm.doc.taxes_and_charges_added = total_gst;
    frm.doc.base_taxes_and_charges_added = total_gst;

    frm.doc.grand_total = final_total;
    frm.doc.base_grand_total = final_total;

    frm.doc.rounded_total = Math.round(final_total);
    frm.doc.base_rounded_total = Math.round(final_total);
    frm.doc.outstanding_amount = final_total;

    frm.refresh_fields([
        "net_total",
        "total_taxes_and_charges",
        "grand_total",
        "rounded_total",
        "discount_amount"
    ]);
}